src/storage/src/source/mod.rs

```diff
@@ -543,16 +557,40 @@ impl fmt::Debug for SourceMessage<Option<Vec<u8>>, Option<Vec<u8>>, ()> {
     }
 }
 
+/// A batch of messages from a source reader, along with the current upper and
+/// any errors that occured while reading that batch.
+pub struct SourceMessageBatch<Key, Value, Diff> {
+    messages: HashMap<PartitionId, Vec<(SourceMessage<Key, Value, Diff>, MzOffset)>>,
+    /// Any errors that occured while obtaining this batch.
+    // TODO: These non-definite errors should not show up in the dataflows/the
+    // persist shard but it's the current "correct" behaviour. We need to fix
+    // this as a follow-up issue because it's a bigger thing that breaks with
+    // the current behaviour.
+    non_definite_errors: Vec<SourceError>,
+    /// The current upper of the `SourceReader`, at the time this batch was
+    /// emitted. Source uppers of emitted batches must never regress.
+    source_upper: HashMap<PartitionId, MzOffset>,
+}
+
+/// The source upper at the time of emitting a batch. This contains only the
+/// partitions that a given source reader operator is responsible for, so a
```

Comment:
It is true, because the `remap_operator` does not yet do proper "mutable frontier" accounting the way you suggested. So if one operator would emit a `(pid, MAX)` here, that would "drop" that pid in the remap operator.